README.md
![](screenshot.png)

SpeedMeter is a versatile Kivy widget aimed at representing numerical
‘physical’ values in a “clock-face” or dial manner.

Minimum and maximum value, and the start and end angle, notably, are
fully configurable. For the sake of readability, the “labels”
(intermediate values represented on the dial) are integers, but you
can override the display function to represent whatever string fits
your needs, as illustrated with PI fractions in demo1.

SpeedMeter is clickable, so you can click on it, and the widget is
able to compute the corresponding physical value.

All default values are “reasonable”, so if you do nothing, you get a
working widget for the range 0..100 (asymmetric, because I think it’s
cool).

For now, SpeedMeter has been tested under Linux and Android (with
buildozer).

Speedmeter functionalities were heavily inspired by Andrea Gavana’s
SpeedMeter for wxPython.

The documentation is at https://kivy-garden.github.io/speedmeter/


CHANGELOG
0.1.2 2020/01/01
================
	* Fixed setup.py so images (necessary for demos to work) are installed
	* fixed setup.py so demos are installed (in share/kivy_garden/speedmeter)

0.1.0 Initial release
======================


demos/demo1.py
from kivy.config import Config
Config.set('input', 'mouse', 'mouse,multitouch_on_demand')

from math import exp

from kivy.app import App
from kivy.clock import Clock
from kivy.factory import Factory

from kivy_garden.speedmeter import SpeedMeter

class NoValueSpeedMeter(SpeedMeter):

    def value_str(self, n): return ''

_displayed = { 
    0: '0',
    30: u'\u03a0 / 6', 60: u'\u03a0/3', 90: u'\u03a0/2', 120: u'2\u03a0/3',
    150: u'5\u03a0/6',
    180: u'\u03a0', 210: u'7\u03a0/6', 240: u'4\u03a0/3'
    }

class PiValueSpeedMeter(SpeedMeter):

    def value_str(self, n):
        return _displayed.get(n, '')

class Demo1(App):

    def __init__(self):
        App.__init__(self)
        self.clockRunning = True

    def tick(self, *t, **kw):
        ids = self.root.ids

        if self.clockRunning:
            clock = ids.clock
            clock.value += 0.2
            if clock.value > clock.max - 0.001: clock.value = clock.min

        fuel = ids.fuel
        fuel.value -= 0.02 * ids.rpm.value
        if fuel.value < fuel.min: fuel.value = fuel.min

        speed_value = ids.speed_value
        if fuel.value <= fuel.min and speed_value.value > speed_value.min:
            speed_value.value = speed_value.value * 0.9 - 3
            if speed_value.value <= speed_value.min:
                speed_value.value = speed_value.min
                speed_value.disabled = True
        
    def set_speed(self):
        ids = self.root.ids
        fuel = ids.fuel
        ids.rpm.value = exp(ids.speed_value.value / 200.0) * 4.5 - 4.5

    def set_temp(self, temperature, motion):
        v = temperature.get_value(motion.pos)
        if not v: return
        temperature.value = v
        self.root.ids.temperature_display.text = '%.2f' % v

    def set_pi(self):
        ids = self.root.ids
        try: v = float(self.root.ids.pi_text.text)
        except: return
        if v < 0 or v > 240: return
        ids.pi.value = v

example = Demo1()
Clock.schedule_interval(example.tick, 0.5)
example.run()


demos/demo2.py
from kivy.config import Config
Config.set('graphics', 'width', 1000)
Config.set('graphics', 'height', 775)
Config.set('input', 'mouse', 'mouse,multitouch_on_demand')

from kivy.app import App
from kivy.factory import Factory
from kivy.properties import *

from kivy_garden.speedmeter import SpeedMeter

class Demo2(App):

    currentColor = StringProperty('')

    def on_start(self):
        self.sm = self.root.ids['sm']
        self.sm.bind(on_touch_down=self._touch_down)
        self.field2Set = False # Dirty

    def _setVectorColor(self, hex_color):
        sm = self.sm
        if not sm.sectors:
            sm.sectors = [sm.min]
        if self.lastPick <= sm.sectors[-1]: return
        sm.sectors = sm.sectors + [hex_color, self.lastPick]
        self.root.ids['sector_width'].disabled = False

    def _setCadranColor(self, hex_color):
        self.sm.cadran_color = hex_color

    def _setShadowColor(self, hex_color):
        self.sm.shadow_color = hex_color

    def _setNeedleColor(self, hex_color):
        self.sm.needle_color = hex_color

    def _touch_down(self, sm, motionEvent):
        if motionEvent.button == 'left':
            v = sm.get_value(motionEvent.pos)
            if v is None: return
            sm.value = v
        else:
            self.lastPick = sm.get_value(motionEvent.pos)
            if self.sm.sectors and self.lastPick <= self.sm.sectors[-1]: return
            self.colorPicker = Factory.ColorPickerDialog()
            self.colorPicker._exampleCallback = self._setVectorColor
            self.colorPicker.open()

    def setLabelControlState(self, modal):
        #if modal.field2Set != 'label_icon_scale': return
        w = self.root.ids
        w['label_icon_scale'].disabled = not self.sm.label_icon
        sm = self.sm
        disabled = not sm.label and not sm.label_icon
        w['label_radius_ratio'].disabled = disabled
        w['label_angle_ratio'].disabled = disabled

        disabled = not sm.label or sm.label_icon
        w['label_font_size'].disabled = disabled

Demo2().run()


searchindex.js
Search.setIndex({docnames:["api","demos","history","index","todo","wxpython_differences"],envversion:52,filenames:["api.rst","demos.rst","history.rst","index.rst","todo.rst","wxpython_differences.rst"],objects:{"":{SpeedMeter:[0,0,1,""],collide_point:[0,1,1,""],display_first:[0,2,1,""],display_last:[0,2,1,""],end_angle:[0,2,1,""],get_value:[0,1,1,""],label:[0,2,1,""],label_angle_ratio:[0,2,1,""],label_font_size:[0,2,1,""],label_icon:[0,2,1,""],label_icon_scale:[0,2,1,""],label_radius_ratio:[0,2,1,""],max:[0,2,1,""],min:[0,2,1,""],needle_color:[0,2,1,""],needle_image:[0,2,1,""],sector_width:[0,2,1,""],shadow_color:[0,2,1,""],start_angle:[0,2,1,""],subtick:[0,2,1,""],tick:[0,2,1,""],value:[0,2,1,""],value_font_size:[0,2,1,""],value_str:[0,1,1,""]}},objnames:{"0":["py","class","Python class"],"1":["py","method","Python method"],"2":["py","attribute","Python attribute"]},objtypes:{"0":"py:class","1":"py:method","2":"py:attribute"},terms:{"00ff00":0,"6bf2ff":0,"90s":2,"case":0,"catch":0,"class":0,"default":[0,1,3],"final":2,"float":0,"function":[0,3],"int":0,"return":0,"switch":2,"true":0,For:[0,3,5],The:[0,4],There:0,These:0,Useful:0,abil:2,abl:[2,3],about:0,accomplish:1,act:4,actual:0,after:0,aim:3,all:[1,3],almost:1,along:2,altern:0,andrea:[2,3],android:[2,3],angl:[0,3],api:3,appli:0,arc:0,asymmetr:3,atla:0,back:2,balanc:0,base:2,basi:0,basic:1,becaus:3,been:[2,3],behavior:0,behaviour:0,belov:2,below:0,better:2,between:0,bodi:0,booleanproperti:0,border:0,both:[0,2],buildoz:3,cadran:4,call:0,can:[0,1,3],center:[0,5],central:4,chang:[0,1],circl:0,click:3,clickabl:3,clock:[0,3],clockwis:0,close:0,clutter:0,coher:0,collide_point:0,color:[0,5],come:0,common:0,complet:2,comput:3,configur:3,consol:2,content:3,control:[0,1],conveni:0,convert:0,cool:[0,3],correspond:[0,3],could:0,counter:0,cover:0,current:[0,2],data:[0,2],death:0,decid:2,declar:1,demo1:[0,3],demo2:[0,3],demo:3,deriv:0,describ:0,descript:0,deserv:0,devic:2,dial:[3,5],differ:[2,3],direct:0,displai:[0,3,4],display_first:0,display_last:0,distanc:0,divid:0,divis:0,doesn:0,doing:2,don:0,down:0,draw:0,drawn:0,driven:0,each:0,effect:1,end:[0,3],end_angl:0,enough:[0,1],equal:0,evenli:0,exampl:0,excel:2,exist:0,explanatori:0,express:0,face:[0,3],far:2,feel:0,few:2,ff0000:0,ffffff:0,field:2,file:0,find:0,fit:3,flexibl:2,fraction:[0,3],frame:0,from:[0,5],full:0,fulli:3,gaug:2,gavana:[2,3],geometr:0,get:[0,3],get_valu:0,good:1,gradient:5,graphic:2,greater:0,green:0,guid:1,had:2,half:0,hand:[0,4],happi:2,has:[0,1,2,3],have:0,heavili:3,hex_color:0,histori:3,how:0,howev:2,human:0,hypothermia:0,icon:[0,4],illustr:3,imag:0,implement:0,inclin:0,includ:0,increas:0,index:3,influenc:0,inner:2,insid:0,inspir:3,integ:[0,3],intermedi:3,interpret:0,interv:0,isn:0,its:2,just:2,kind:2,kivi:[0,2,3,4],label:[0,3,4],label_angle_ratio:0,label_font_s:0,label_icon:0,label_icon_scal:0,label_radius_ratio:0,last:0,layout:4,learn:2,less:0,liber:0,lightli:2,like:0,linux:[0,2,3],list:0,listproperti:0,lost:2,lot:2,lower:0,mainli:0,mani:2,manner:3,map:0,mark:0,max:0,maximum:[0,3],mean:[0,2,4],method:3,min:0,minimum:[0,3],mission:1,model:2,monitor:2,more:[0,4],most:0,mostli:0,motionev:0,motor:2,move:0,must:[0,2],name:0,need:[0,2,3],needl:0,needle2:0,needle3:0,needle_color:0,needle_imag:0,neg:0,none:0,notabl:3,note:0,noth:[0,3],now:[3,5],number:0,numer:3,numericproperti:0,occur:0,omit:0,onli:0,opengl:2,opposit:0,order:0,origin:[0,2],other:[2,3],otherwis:0,out:2,outsid:[0,5],over:0,overheat:0,overrid:[0,3],own:0,page:3,pair:0,paramet:[0,1,2],part:0,particular:2,physic:3,pictur:0,place:0,plai:[0,1],platform:3,png:0,point:0,pos:0,posit:0,preced:0,prefer:0,present:0,prevent:0,project:2,properti:0,purpos:0,put:0,quit:0,radiu:0,rang:[0,2,3],ratio:0,readabl:[0,3],real:[0,4],reason:[0,3],red:0,reduc:0,relat:2,replic:1,repres:[0,2,3],represent:3,result:0,rewrit:2,rewritten:1,rgb:0,right:0,robot:2,rotat:[4,5],rpm:2,rule:2,run:[0,2],sai:2,sake:3,same:0,satisfactori:2,saw:2,search:3,sector:3,sector_width:0,see:[0,1],self:0,set:[0,1],shadow:0,shadow_color:0,should:4,silicon:2,similar:0,sinc:0,singl:0,size:0,slider:0,some:0,sort:0,speak:2,special:0,specifi:0,speedmet:[0,1,2,4],standard:0,start:[0,1,2,3],start_angl:0,stolen:0,str:0,string:[0,3],stringproperti:0,sub:0,subtick:0,suit:[0,2],support:5,surfac:0,symbol:0,symmetri:0,tablet:2,take:0,tediou:0,temperatur:0,term:0,territori:2,test:3,text:[0,4],than:0,thei:0,thi:[0,1,2,3],thing:0,think:[2,3],thu:0,tick:0,todo:3,too:0,took:2,travel:2,troubleshoot:2,undecid:2,undefin:0,under:[2,3],unknown:2,upper:0,use:[0,2],used:2,useful:0,using:[0,2],valu:[3,4,5],value_font_s:0,value_str:0,variou:2,veri:2,voltag:2,wai:[0,2],want:0,warm:0,wasn:2,week:2,were:3,what:0,whatev:3,when:[1,2],where:0,which:[0,2],white:0,whole:0,widget:[0,1,3],width:0,without:0,won:0,work:[0,3],workstat:2,would:2,write:[1,2],written:0,wxpython:[1,2,3],yet:0,you:[0,1,3],your:[0,3]},titles:["API","Demos","History","Welcome to SpeedMeter\u2019s documentation!","TODO","Notable differences with wxPython SpeedMeter"],titleterms:{api:0,demo1:1,demo2:1,demo:1,dial:0,differ:5,document:3,histori:2,indic:[0,3],method:0,notabl:5,other:0,physic:0,represent:0,sector:0,speedmet:[3,5],tabl:3,todo:4,valu:0,welcom:3,wxpython:5}})

kivy_garden/speedmeter/__init__.py
"""
SpeedMeter
==========

Defines the Kivy garden :class:`SpeedMeter` class, a versatile
gauge widget inspired by Andrea Gavana's SpeedMeter for wxPython,
which allows display and input of physical values.
"""

import os.path, sys

from math import atan2, cos, pi, radians, sin

from kivy.core.text import Label
from kivy.graphics import *
from kivy.properties import *
from kivy.uix.image import CoreImage, Image
from kivy.uix.widget import Widget
from kivy.utils import get_color_from_hex

__all__ = ('SpeedMeter', )

__version__ = '0.1.2'

_redraw = tuple('pos size min max'.split())
_redraw_background = tuple('sectors sector_width shadow_color'.split())
_redraw_full_cadran = tuple(
    'tick subtick cadran_color display_first '
    'display_last value_font_size'.split())
_redraw_label = tuple(
    'label label_radius_ratio label_angle_ratio '
    'label_icon label_icon_scale label_font_size'.split())
_redraw_needle = tuple('needle_color'.split())

_image_dir = os.path.join(os.path.dirname(__file__), 'images')

from kivy.graphics.instructions import *

_2pi = 2 * pi  # I trust multiplication by 2 even for floating points !
_halfPi = pi / 2

#
# Each part is stores in its own InstructionGroup, so as to minimize
# redrawing / recomputation when something is modified, except
# outerCadran and values that is really one function, but is split for
# the sake of understandability.
#

_ig = tuple('sectors shadow outerCadran values label needle'.split())


class SpeedMeter(Widget):

    """A versatile gauge widget which allows display and input of
    "physical" values."""

    min = NumericProperty(0)
    max = NumericProperty(100)

    tick = NumericProperty(10)
    subtick = NumericProperty(0)

    display_first = BooleanProperty(True)
    display_last = BooleanProperty(True)

    start_angle = NumericProperty(-90, min=-360, max=360)
    end_angle = NumericProperty(135, min=-360, max=360)

    cadran_color = StringProperty('#ffffff')

    label = StringProperty('')
    label_icon = StringProperty('')
    label_icon_scale = NumericProperty(0.5, min=0, max=1)
    label_font_size = NumericProperty(15, min=1)
    value_font_size = NumericProperty(15, min=1)

    label_radius_ratio = NumericProperty(0.3, min=-1, max=1)
    label_angle_ratio = NumericProperty(0.5, min=0, max=1)

    needle_color = StringProperty('#6bf2ff')
    needle_image = StringProperty('needle.png')

    sectors = ListProperty()
    sector_width = NumericProperty(0, min=0)
    thickness = NumericProperty(1.5, min=0)

    shadow_color = StringProperty('')

    value = NumericProperty(0)

    def __init__(self, **kwargs):
        super(SpeedMeter, self).__init__(**kwargs)
        # In case on_value is called before _update
        self.a = self.b = self.r2 = self.r = self.centerx = self.centery = 0
        self._shadow = None
        self.rotate = _X
        self._labelRectangle = -1
        self._needle_image = os.path.join(_image_dir, 'needle.png')

        add = self.canvas.add
        for instructionGroupName in _ig:
            ig = InstructionGroup()
            setattr(self, '_%sIG' % instructionGroupName, ig)
            add(ig)

        self.extendedTouch = False
        bind = self.bind
        for eventList, fn in (
                (_redraw, self._redraw),
                (_redraw_background, self._draw_background),
                (_redraw_full_cadran, self._draw_full_cadran),
                (_redraw_label, self._draw_label),
                (_redraw_needle, self._draw_needle),
        ):
            for event in eventList:
                bind(**{event: fn})

    def value_str(self, n):
        """This is the method called to convert data n into the string
represented on the dial. It defaults to str(int(n)). If you want to
have more control over what's written on the dial, you can derive your
own class from SpeedMeter and override this function. See demo1 for an
example using fractions of PI."""
        # Override this if you want more control on the tick display
        return str(int(n))

    def _draw_sectors(self):
        self._sectorsIG.clear()
        add = self._sectorsIG.add
        l = self.sectors[:]
        if not l:
            return
        r = self.r
        centerx = self.centerx
        centery = self.centery
        d = r + r
        a = self.a
        b = self.b
        v0 = l.pop(0)
        a0 = -(a * v0 + b)
        sw = self.sector_width
        if sw:
            r -= sw
        else:
            centerx -= r
            centery -= r
            dd = (d, d)

        while l:
            color = l.pop(0)
            v1 = l.pop(0) if l else self.max
            a1 = -(a * v1 + b)
            add(Color(rgba=get_color_from_hex(color)))
            if sw:
                add(Line(circle=(centerx, centery, r, a0, a1),
                         width=sw, cap='none'))
            else:
                add(Ellipse(pos=(centerx, centery),
                            size=dd, angle_start=a0, angle_end=a1))
            a0 = a1

    def _set_shadow_value(self):
        if not self._shadow:
            return
        a0 = -(self.a * self.min + self.b)
        a1 = -(self.a * self.value + self.b)
        self._shadow.circle = (self.centerx, self.centery, self.r - 5, a0, a1)

    def _draw_shadow(self):
        self._shadowIG.clear()
        self._shadow = None
        if not self.shadow_color:
            return
        add = self._shadowIG.add
        add(Color(rgba=get_color_from_hex(self.shadow_color)))
        self._shadow = Line(width=5, cap='none')
        add(self._shadow)
        self._set_shadow_value()

    def _draw_outer_cadran(self):
        self._outerCadranIG.clear()
        add = self._outerCadranIG.add
        centerx = self.centerx
        centery = self.centery
        r = self.r
        theta0 = self.start_angle
        theta1 = self.end_angle
        add(Color(rgba=get_color_from_hex(self.cadran_color)))
        if theta0 == theta1:
            add(Line(circle=(centerx, centery, r), width=1.5))
        else:
            rt0 = radians(theta0)
            rt1 = radians(theta1)
            add(Line(points=(
                centerx + r * sin(rt0),
                centery + r * cos(rt0),
                centerx, centery,
                centerx + r * sin(rt1),
                centery + r * cos(rt1),
            ),
                     width=1.5,
            ))
            add(Line(circle=(centerx, centery, r, theta0, theta1), width=1.5))

    # I'm using theta for the angle, as to not confuse it with
    # transparency (alpha as in in rgba)

    def _draw_values(self):
        self._valuesIG.clear()
        add = self._valuesIG.add
        centerx = self.centerx
        centery = self.centery
        r = self.r
        value_str = self.value_str
        values = [
            Label(value_str(i), bold=True, font_size=self.value_font_size)
            for i in range(self.min, self.max + 1, self.tick)]
        if len(values) <= 1:
            # Tick is bigger than max - min
            return
        for _ in values:
            _.refresh()

        theta0 = self.start_angle
        theta1 = self.end_angle
        if theta0 == theta1:
            theta1 += 360
        deltaTheta = radians((theta1 - theta0) / float(len(values) - 1))
        theta = radians(theta0)
        r_10 = r - 10
        r_20 = r - 20
        subtick = int(self.subtick)
        if subtick:
            subDeltaTheta = deltaTheta / subtick
        else:
            subDeltaTheta = None
        # Should be more efficient on texture allocation
        for value in values:
            first = value is values[0]
            last = value is values[-1]
            c = cos(theta)
            s = sin(theta)
            r_1 = r - 1
            if (
                    not first and not last
                    or first and self.display_first
                    or last and self.display_last):
                # Draw the big tick
                add(Line(points=(
                    centerx + r_1 * s, centery + r_1 * c,
                    centerx + r_10 * s, centery + r_10 * c,
                ), width=2))
                # Numerical value
                t = value.texture
                tw, th = t.size
                add(Rectangle(
                    pos=(centerx + r_20 * s - tw / 2,
                         centery + r_20 * c - th / 2),
                    size=t.size,
                    texture=t))
            # Subtick
            if subDeltaTheta and not last:
                subTheta = theta + subDeltaTheta
                for n in range(subtick):
                    subc = cos(subTheta)
                    subs = sin(subTheta)
                    add(Line(points=(
                        centerx + r * subs, centery + r * subc,
                        centerx + r_10 * subs, centery + r_10 * subc),
                             width=0.75))
                    subTheta += subDeltaTheta
            theta += deltaTheta

    def _draw_label(self, *t):
        self._labelIG.clear()
        if not self.label and not self.label_icon:
            return

        theta = self.start_angle + (
            self.label_angle_ratio * (self.end_angle - self.start_angle))
        c = cos(radians(theta))
        s = sin(radians(theta))
        r = self.r
        r1 = r * self.label_radius_ratio
        if self.label_icon:
            label = CoreImage(self.label_icon)
            t = label.texture
            iconSize = max(t.size)
            scale = r * self.label_icon_scale / float(iconSize)
            tw, th = t.size
            tw *= scale
            th *= scale
        else:
            label = Label(text=self.label, markup=True, bold=True,
                          font_size=self.label_font_size)
            label.refresh()
            t = label.texture
            tw, th = t.size
        self._labelIG.add(
            Rectangle(
                pos=(self.centerx + r1 * s - tw / 2,
                     self.centery + r1 * c - th / 2),
                size=(tw, th),
                texture=t))

    def on_needle_image(self, *t):
        self._needle_image = self.needle_image
        full_path = os.path.join(_image_dir, self._needle_image)
        if os.path.isfile(full_path):
            self._needle_image = full_path
        self._draw_needle(*t)

    def _draw_needle(self, *t):
        self._needleIG.clear()
        add = self._needleIG.add
        add(PushMatrix())
        self.rotate = Rotate(origin=(self.centerx, self.centery))
        add(self.rotate)

        if self.value < self.min:
            self.value = self.min
        elif self.value > self.max:
            self.value = self.max
        needleSize = self.r
        s = needleSize * 2
        add(Color(rgba=get_color_from_hex(self.needle_color)))
        add(Rectangle(
            pos=(self.centerx - needleSize, self.centery - needleSize),
            size=(s, s),
            source=self._needle_image))
        add(PopMatrix())
        self.on_value()

    def _draw_background(self, *t):
        self._draw_sectors()
        self._draw_shadow()

    def _draw_full_cadran(self, *t):
        self._draw_outer_cadran()
        self._draw_values()

    def _redraw(self, *args):
        diameter = min(self.size)
        sa = self.start_angle
        ea = self.end_angle

        r = self.r = diameter / 2
        self.r2 = r * r
        x, y = self.pos
        width, height = self.size
        self.centerx = x + width / 2
        self.centery = y + height / 2

        #
        # compute value -> angle mapping
        #

        theta0 = sa
        theta1 = ea
        if theta0 == theta1:
            theta1 += 360
        self.a = (float(theta0) - theta1) / (self.max - self.min)
        self.b = -theta0 - self.a * self.min

        #
        # Reverse mapping
        #
        self.startTheta = _halfPi - radians(sa)
        self.endTheta = _halfPi - radians(ea)
        self.direct = self.startTheta < self.endTheta

        self.ra = (self.max - self.min) / ((self.endTheta - self.startTheta)
                                           if sa != ea else _2pi)
        self.rb = self.min - self.ra * self.startTheta

        #
        # Draw
        #
        self._draw_background()
        self._draw_full_cadran()
        self._draw_label()
        self._draw_needle()

    def on_value(self, *t):
        self.rotate.angle = self.a * self.value + self.b
        self._set_shadow_value()

    def get_value(self, pos):
        c = self.center
        x = pos[0] - c[0]
        y = pos[1] - c[1]
        r2 = x * x + y * y
        if r2 > self.r2:
            return
        theta = atan2(y, x)
        theta_ = theta

        min_, max_ = ((self.startTheta, self.endTheta)
                      if self.direct
                      else (self.endTheta, self.startTheta))
        if theta < min_:
            theta += _2pi
        elif theta > max_:
            theta -= _2pi

        v = self.ra * theta + self.rb
        if v >= self.min and v <= self.max:
            return v
        if not self.extendedTouch:
            return
        # Should make distinction between min and max here

    def collide_point(self, x, y):
        return self.get_value(*(x, y)) is not None

    def on_start_angle(self, *t):
        if self.end_angle - self.start_angle > 360:
            self.start_angle = self.end_angle - 360
        self._redraw()

    def on_end_angle(self, *t):
        if self.end_angle - self.start_angle > 360:
            self.end_angle = self.start_angle + 360
        self._redraw()


class _X:
    pass
